plonk-hashing/src/poseidon/poseidon_ref.rs

```diff
@@ -0,0 +1,525 @@
+        v: Self::ParameterField,
+    ) -> Self::Field {
+        c.add_input(v)
+    }
+
+    fn zeros<const W: usize>(
+        c: &mut StandardComposer<F, P>,
+    ) -> [Self::Field; W] {
+        [c.zero_var(); W]
+    }
+
+    fn add(
+        c: &mut StandardComposer<F, P>,
+        x: &Self::Field,
+        y: &Self::Field,
+    ) -> Self::Field {
+        c.arithmetic_gate(|g| g.witness(*x, *y, None).add(F::one(), F::one()))
+    }
+
+    fn addi(
+        c: &mut StandardComposer<F, P>,
+        a: &Self::Field,
+        b: &Self::ParameterField,
+    ) -> Self::Field {
+        let zero = c.zero_var();
+        c.arithmetic_gate(|g| {
+            g.witness(*a, zero, None)
+                .add(F::one(), F::zero())
+                .constant(*b)
+        })
+    }
+
+    fn mul(
+        c: &mut StandardComposer<F, P>,
+        x: &Self::Field,
+        y: &Self::Field,
+    ) -> Self::Field {
+        c.arithmetic_gate(|q| q.witness(*x, *y, None).mul(F::one()))
+    }
+
+    fn muli(
+        c: &mut StandardComposer<F, P>,
+        x: &Self::Field,
+        y: &Self::ParameterField,
+    ) -> Self::Field {
+        let zero = c.zero_var();
+        c.arithmetic_gate(|g| g.witness(*x, zero, None).add(*y, F::zero()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ark_ec::PairingEngine;
+    type E = ark_bls12_381::Bls12_381;
+    type P = ark_ed_on_bls12_381::EdwardsParameters;
+    type Fr = <E as PairingEngine>::Fr;
+    use ark_ff::Field;
+    use ark_std::{test_rng, UniformRand};
+
+    #[test]
+    // poseidon should output something if num_inputs = arity
+    fn sanity_test() {
+        const ARITY: usize = 4;
+        const WIDTH: usize = ARITY + 1;
+        let mut rng = test_rng();
+
+        let param = PoseidonConstants::generate::<WIDTH>();
+        let mut poseidon = PoseidonRef::<(), NativeSpecRef<Fr>, WIDTH>::new(
+            &mut (),
+            param.clone(),
+        );
+        let inputs = (0..ARITY).map(|_| Fr::rand(&mut rng)).collect::<Vec<_>>();
+
+        inputs.iter().for_each(|x| {
+            let _ = poseidon.input(*x).unwrap();
+        });
+        let native_hash: Fr = poseidon.output_hash(&mut ());
+
+        let mut c = StandardComposer::<Fr, P>::new();
+        let inputs_var =
+            inputs.iter().map(|x| c.add_input(*x)).collect::<Vec<_>>();
+        let mut poseidon_circuit =
+            PoseidonRef::<_, PlonkSpecRef, WIDTH>::new(&mut c, param);
+        inputs_var.iter().for_each(|x| {
+            let _ = poseidon_circuit.input(*x).unwrap();
+        });
+        let plonk_hash = poseidon_circuit.output_hash(&mut c);
+
+        c.check_circuit_satisfied();
+
+        let expected = c.add_input(native_hash);
+        c.assert_equal(expected, plonk_hash);
+
+        c.check_circuit_satisfied();
+        println!(
+            "circuit size for WIDTH {} poseidon: {}",
+            WIDTH,
+            c.circuit_size()
+        )
+    }
+
+    #[test]
```

Comment:
we do not need to test on `r1cs`. 

(Indeed, we should remove everything related `r1cs` here. It's not relevant. )